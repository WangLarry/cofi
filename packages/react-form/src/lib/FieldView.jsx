/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import PropTypes from 'prop-types';
import noop from 'lodash/noop';
import * as Styled from './Styled';

const propTypes = {
  id: PropTypes.string.isRequired,
  label: PropTypes.string,
  description: PropTypes.string,
  component: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({ render: PropTypes.func.isRequired }),
  ]),
  value: PropTypes.any,
  state: PropTypes.object,
  excluded: PropTypes.bool,
  disabled: PropTypes.bool,
  dirty: PropTypes.bool,
  required: PropTypes.bool,
  empty: PropTypes.bool,
  invalid: PropTypes.bool,
  errors: PropTypes.array,
  styleOverrides: PropTypes.string,
  onValueChange: PropTypes.func,
  onStateChange: PropTypes.func,
};

const FieldView = ({
  id = undefined,
  label = undefined,
  description = undefined,
  value = undefined,
  excluded = false,
  disabled = false,
  dirty = false,
  required = false,
  empty = false,
  invalid = false,
  errors = [],
  state = {},
  component = undefined,
  styleOverrides = undefined,
  onValueChange = noop,
  onStateChange = noop,
  ...customProps
}) => {
  if (excluded) return null;

  const Component = component;

  return (
    <Styled.Field 
      id={id}
      aria-label="Field" 
      data-disabled={disabled} 
      data-dirty={dirty} 
      data-required={required}
      data-empty={empty} 
      data-invalid={invalid}
      styleOverrides={styleOverrides}>
      {
        label && <div aria-label="Header">
          <div aria-label="Label">{label}</div>
          { required && (<div aria-label="Required">*</div>) }
          { description && (<i aria-label="Description" title={description}>?</i>) }
        </div>
      }
      {
        Component && <div aria-label="Component">
          <Component
            {...customProps}
            value={value}
            state={state}
            disabled={disabled}
            dirty={dirty}
            required={required}
            empty={empty}
            invalid={invalid} 
            onValueChange={onValueChange}
            onStateChange={onStateChange} />
        </div>
      }
      {
        errors.length > 0 && <div aria-label="Error">{errors[0].message}</div>
      }
    </Styled.Field>
  );
};

FieldView.prototype = propTypes;

export default FieldView;
