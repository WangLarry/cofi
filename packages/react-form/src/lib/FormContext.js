/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';

const FormContext = React.createContext();
export default FormContext;
