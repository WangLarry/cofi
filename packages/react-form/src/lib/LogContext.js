/**
  * Copyright 2020, Gal Havivi
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';

const LogContext = React.createContext();
export default LogContext;
