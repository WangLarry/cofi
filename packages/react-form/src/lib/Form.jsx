/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React, { useState, useEffect, useContext } from 'react';
import PropTypes from 'prop-types';
import JForm from '@cofi/form';
import FormContext from './FormContext';

const Form = ({ model, resources, settings, data, context, children }) => {
  const [form, setForm] = useState();
  const [privates] = useState({ models: [] });
  const parent = useContext(FormContext);

  const safeDestroy = async () => {
    // if last init was a success (not failed on definition error)
    if (privates.form && !privates.form.destroyed && privates.lastInitSuccess) {
      await privates.form.destroy();
    }
  };

  // on init and on model prop changes - init
  // queue the changed models, for cases models was changed few times in a row without finish its first init
  useEffect(() => {
    const updatePublicForm = (form) => !privates.destroyed && setForm(form);

    const init = async () => {
      privates.models.push(model);

      if (!privates.processing) {
        privates.processing = true;

        while (privates.models.length) {
          await safeDestroy();
          privates.form = new JForm();
          const currModel = privates.models[0];
          const contextOverride = context ? { context: context } : undefined;
          const formModel = Object.assign({}, currModel, contextOverride);
          privates.lastInitSuccess = await privates.form.init(formModel, resources, settings, updatePublicForm);
          privates.models.shift();
        }
      
        privates.processing = false;
      }
    };

    init();
  }, [model]); // eslint-disable-line

  // on un-mount - mark destroyed, to disable upcoming ui dispatches
  useEffect(() => {
    return () => {
      safeDestroy();
      privates.destroyed = true; 
    };
  }, []); // eslint-disable-line

  // when data prop changes - call changeData
  useEffect(() => { 
    privates.firstTimeData && privates.form.changeData(data);
    privates.firstTimeData = true;
  }, [data]); // eslint-disable-line

  // when context prop changes - call changeData
  useEffect(() => {
    privates.firstTimeContext && privates.form.changeContext(context);
    privates.firstTimeContext = true;
  }, [context]); // eslint-disable-line

  if (!form) return (null);

  const contextValue = {
    parent,
    model: form.model,
    resources: form.resources,
    actions: {
      changeValue: privates.form.changeValue.bind(privates.form),
      changeData: privates.form.changeData.bind(privates.form),
      changeContext: privates.form.changeContext.bind(privates.form),
      changeState: privates.form.changeState.bind(privates.form),
      changeUi: privates.form.changeUi.bind(privates.form),
      submit: privates.form.submit.bind(privates.form),
      reset: privates.form.reset.bind(privates.form),
    },
  };

  return (<FormContext.Provider value={contextValue}><>{children}</></FormContext.Provider>);
};

Form.prototype = {
  model: PropTypes.object.isRequired,
  resources: PropTypes.object,
  settings: PropTypes.object,
  data: PropTypes.object,
  context: PropTypes.object,
};

export default Form;
