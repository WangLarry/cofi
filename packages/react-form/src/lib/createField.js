/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React, { useState, useContext } from 'react';
import PropTypes from 'prop-types';
import isEqual from 'lodash/isEqual';
import { mapFieldToProps } from '@cofi/form';
import FormContext from './FormContext';

export default function createField(FieldView) {

  const Field = ({ id, ...customProps }) => {
    const [cachedProps, setCachedProps] = useState();
    const [cachedFieldView, setCachedFieldView] = useState((null));
    const { model, resources, actions } = useContext(FormContext);
    
    const props = mapFieldToProps(id, model, resources);
  
    // optimization prevent un-necessary field renders
    if (!isEqual(props, cachedProps)) {
      setCachedProps(props);
      setCachedFieldView(<FieldView
        {...customProps}
        {...props}
        onValueChange={value => actions.changeValue(id, value)}
        onStateChange={state => actions.changeState(id, state)}
      />);
    }

    return cachedFieldView;
  };
  
  Field.propTypes = {
    id: PropTypes.string.isRequired,
  };

  return Field;
}
