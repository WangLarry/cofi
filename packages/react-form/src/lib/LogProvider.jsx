/**
  * Copyright 2020, Gal Havivi
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React, { useState, useEffect, useCallback } from 'react';
import cloneDeep from 'lodash/cloneDeep';
import { setLogger, Steps } from '@cofi/form';
import LogContext from './LogContext';

const defaultSettings = { 
  debugMaxLength: 50, 
  errorMaxLength: 10, 
  warnMaxLength: 10,
  formIds: [],
};

const initialHistory = { debug: [], warn: [], error: [] };

const addRecord = (history, funcName, record, maxLength) => {
  history[funcName].push(record);

  if (history[funcName].length > maxLength) {
    history[funcName].splice(0, 1);
  }
};

const addRecordAndUpdateLog = (history, setHistory, funcName, record, maxLength, shouldSet = () => true) => {
  addRecord(history, funcName, record, maxLength);

  // optimization - react render for historys component only on specific condition
  if (shouldSet()) {
    setTimeout(() => { // optimization - make page loads faster, since debug message called during init action
      setHistory({ 
        debug: history.debug, 
        warn: history.warn,
        error: history.error,
      });
    });
  }
};

// step exists only for debug records
const getRecordFormId = ({ form, step = {} }) => form && form.model ? form.model.id : step.formId;

const shouldAdd = (record, settings) => !settings.formIds.length ||
  settings.formIds.includes(getRecordFormId(record));

const customLogger = (history, setHistory, settings) => ({
  error: record => shouldAdd(record, settings)
    && addRecordAndUpdateLog(history, setHistory, 'error', record, settings.errorMaxLength),
  warn: record => shouldAdd(record, settings)
    && addRecordAndUpdateLog(history, setHistory, 'warn', record, settings.warnMaxLength),
  debug: record => shouldAdd(record, settings)
    && addRecordAndUpdateLog(history, setHistory, 'debug', record, settings.debugMaxLength,
      () => record.step.type === Steps.END_PROCESSING),
});

const generateId = () => Math.random().toString(16).slice(2);

export default (props) => {
  const [history, setHistory] = useState(cloneDeep(initialHistory));
  const [settings, setSettings] = useState();

  useEffect(() => {
    const _settings = { ...defaultSettings, ...props.settings };
    const name = generateId(); // unique between multiple log provider
    setLogger(name, customLogger(history, setHistory, _settings));
    setSettings(_settings);

    // cleanup
    return () => setLogger(name);
  }, [history, setHistory, props.settings]);

  const clear = useCallback(() => setHistory(cloneDeep(initialHistory)), [setHistory]);

  if (!settings) return (null);

  const context = {
    settings,
    history,
    clear,
  };

  return (<LogContext.Provider value={context}>{ props.children }</LogContext.Provider>);
};
