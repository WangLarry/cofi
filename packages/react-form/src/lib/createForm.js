/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import Form from './Form';

export default function createForm({ model, resources, settings }) {
  return Inner => props => {
    const { context, data, ...customProps } = props;
    return (<Form model={model} resources={resources} settings={settings} context={context} data={data}>
      <Inner { ...customProps } />
    </Form>);
  };
}
