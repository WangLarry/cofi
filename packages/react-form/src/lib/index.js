/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import FormContext from './FormContext';
import FieldView from './FieldView';
import Field from './Field';
import Form from './Form';
import LogContext from './LogContext';
import LogProvider from './LogProvider';
import createForm from './createForm';
import createField from './createField';
import withForm from './withForm';
import withLog from './withLog';

export {
  FormContext,
  Form,
  Field,
  FieldView,
  LogContext,
  LogProvider,
  createForm,
  createField,
  withForm,
  withLog,
};
