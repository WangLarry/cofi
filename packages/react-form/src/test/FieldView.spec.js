/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import PropTypes from 'prop-types';
import { shallow } from 'enzyme';
import { FieldView } from '../lib';

describe('FieldView', () => {
  let props;
  let element;

  beforeEach(() => {
    props = {
      id: 'firstName',
      label: 'First Name',
      description: 'Enter your first name',
      value: 'Rachel Green',
      state: {},
      excluded: undefined,
      disabled: undefined,
      required: undefined,
      invalid: undefined,
      dirty: undefined,
      empty: undefined,
      errors: undefined,
      component: jest.fn(),
      onValueChange: jest.fn(),
      onStateChange: jest.fn(),
    };
  });

  describe('PropTypes', () => {
    it('component - valid when component is forwardRef object', () => {
      const component = React.forwardRef((props, ref) => React.createElement('div', { ref, ...props }));
      testComponentPropType(component);
    });

    it('component - valid when component is class component', () => {
      class myInput extends React.Component {
        render() { return (<div>{this.name}</div>); }
      }
      testComponentPropType(myInput);
    });

    it('component - valid when component is functional component', () => {
      const component = () => {};
      testComponentPropType(component);
    });

    function testComponentPropType(component, errorsNumber = 0) {
      const { error } = global.console;
      global.console.error = jest.fn();
      const fieldViewProps = Object.assign(props, { component });
      PropTypes.checkPropTypes(FieldView.propTypes, fieldViewProps, 'prop', 'fieldView');
      expect(global.console.error.mock.calls).toHaveLength(errorsNumber);
      global.console.error = error;
    }
  });

  describe('render', () => {
    it('default test props', () => {
      element = shallow(<FieldView {...props} />);
      expect(element).toMatchSnapshot();
    });

    it('id = undefined', () => {
      props.id = undefined;
      element = shallow(<FieldView {...props} />);
      expect(element).toMatchSnapshot();
    });

    it('label = undefined', () => {
      props.label = undefined;
      element = shallow(<FieldView {...props} />);
      expect(element).toMatchSnapshot();
    });

    it('description = undefined', () => {
      props.description = undefined;
      element = shallow(<FieldView {...props} />);
      expect(element).toMatchSnapshot();
    });

    it('value = undefined', () => {
      props.value = undefined;
      element = shallow(<FieldView {...props} />);
      expect(element).toMatchSnapshot();
    });

    it('state = undefined', () => {
      props.state = undefined;
      element = shallow(<FieldView {...props} />);
      expect(element).toMatchSnapshot();
    });
  
    it('excluded = true', () => {
      props.excluded = true;
      element = shallow(<FieldView {...props} />);
      expect(element.type()).toEqual(null);
      expect(element).toMatchSnapshot();
    });
  
    it('disabled = true', () => {
      props.disabled = true;
      element = shallow(<FieldView {...props} />);
      expect(element).toMatchSnapshot();
    });

    it('dirty = true', () => {
      props.dirty = true;
      element = shallow(<FieldView {...props} />);
      expect(element).toMatchSnapshot();
    });

    it('required = true', () => {
      props.required = true;
      element = shallow(<FieldView {...props} />);
      expect(element).toMatchSnapshot();
    });

    it('empty = true', () => {
      props.empty = true;
      element = shallow(<FieldView {...props} />);
      expect(element).toMatchSnapshot();
    });

    it('invalid = true', () => {
      props.invalid = true;
      element = shallow(<FieldView {...props} />);
      expect(element).toMatchSnapshot();
    });

    it('errors = [...]', () => {
      props.errors = [{ name: 'munLength', message: 'min length is 2' }];
      element = shallow(<FieldView {...props} />);
      expect(element).toMatchSnapshot();
    });

    it('component = undefined', () => {
      props.component = undefined;
      element = shallow(<FieldView {...props} />);
      expect(element).toMatchSnapshot();
    });

    it('onValueChange = undefined', () => {
      props.onValueChange = undefined;
      element = shallow(<FieldView {...props} />);
      expect(element).toMatchSnapshot();
    });

    it('onStateChange = undefined', () => {
      props.onStateChange = undefined;
      element = shallow(<FieldView {...props} />);
      expect(element).toMatchSnapshot();
    });
  
    it('un-supported cofi field props - pass it to the underline component', () => {
      element = shallow(<FieldView mickey="mouse" {...props} />);
      expect(element).toMatchSnapshot();
    });
  });
});
