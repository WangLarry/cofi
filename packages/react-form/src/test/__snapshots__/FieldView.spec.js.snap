// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldView render component = undefined 1`] = `
<styled.div
  aria-label="Field"
  data-dirty={false}
  data-disabled={false}
  data-empty={false}
  data-invalid={false}
  data-required={false}
  id="firstName"
>
  <div
    aria-label="Header"
  >
    <div
      aria-label="Label"
    >
      First Name
    </div>
    <i
      aria-label="Description"
      title="Enter your first name"
    >
      ?
    </i>
  </div>
</styled.div>
`;

exports[`FieldView render default test props 1`] = `
<styled.div
  aria-label="Field"
  data-dirty={false}
  data-disabled={false}
  data-empty={false}
  data-invalid={false}
  data-required={false}
  id="firstName"
>
  <div
    aria-label="Header"
  >
    <div
      aria-label="Label"
    >
      First Name
    </div>
    <i
      aria-label="Description"
      title="Enter your first name"
    >
      ?
    </i>
  </div>
  <div
    aria-label="Component"
  >
    <mockConstructor
      dirty={false}
      disabled={false}
      empty={false}
      invalid={false}
      onStateChange={[MockFunction]}
      onValueChange={[MockFunction]}
      required={false}
      state={Object {}}
      value="Rachel Green"
    />
  </div>
</styled.div>
`;

exports[`FieldView render description = undefined 1`] = `
<styled.div
  aria-label="Field"
  data-dirty={false}
  data-disabled={false}
  data-empty={false}
  data-invalid={false}
  data-required={false}
  id="firstName"
>
  <div
    aria-label="Header"
  >
    <div
      aria-label="Label"
    >
      First Name
    </div>
  </div>
  <div
    aria-label="Component"
  >
    <mockConstructor
      dirty={false}
      disabled={false}
      empty={false}
      invalid={false}
      onStateChange={[MockFunction]}
      onValueChange={[MockFunction]}
      required={false}
      state={Object {}}
      value="Rachel Green"
    />
  </div>
</styled.div>
`;

exports[`FieldView render dirty = true 1`] = `
<styled.div
  aria-label="Field"
  data-dirty={true}
  data-disabled={false}
  data-empty={false}
  data-invalid={false}
  data-required={false}
  id="firstName"
>
  <div
    aria-label="Header"
  >
    <div
      aria-label="Label"
    >
      First Name
    </div>
    <i
      aria-label="Description"
      title="Enter your first name"
    >
      ?
    </i>
  </div>
  <div
    aria-label="Component"
  >
    <mockConstructor
      dirty={true}
      disabled={false}
      empty={false}
      invalid={false}
      onStateChange={[MockFunction]}
      onValueChange={[MockFunction]}
      required={false}
      state={Object {}}
      value="Rachel Green"
    />
  </div>
</styled.div>
`;

exports[`FieldView render disabled = true 1`] = `
<styled.div
  aria-label="Field"
  data-dirty={false}
  data-disabled={true}
  data-empty={false}
  data-invalid={false}
  data-required={false}
  id="firstName"
>
  <div
    aria-label="Header"
  >
    <div
      aria-label="Label"
    >
      First Name
    </div>
    <i
      aria-label="Description"
      title="Enter your first name"
    >
      ?
    </i>
  </div>
  <div
    aria-label="Component"
  >
    <mockConstructor
      dirty={false}
      disabled={true}
      empty={false}
      invalid={false}
      onStateChange={[MockFunction]}
      onValueChange={[MockFunction]}
      required={false}
      state={Object {}}
      value="Rachel Green"
    />
  </div>
</styled.div>
`;

exports[`FieldView render empty = true 1`] = `
<styled.div
  aria-label="Field"
  data-dirty={false}
  data-disabled={false}
  data-empty={true}
  data-invalid={false}
  data-required={false}
  id="firstName"
>
  <div
    aria-label="Header"
  >
    <div
      aria-label="Label"
    >
      First Name
    </div>
    <i
      aria-label="Description"
      title="Enter your first name"
    >
      ?
    </i>
  </div>
  <div
    aria-label="Component"
  >
    <mockConstructor
      dirty={false}
      disabled={false}
      empty={true}
      invalid={false}
      onStateChange={[MockFunction]}
      onValueChange={[MockFunction]}
      required={false}
      state={Object {}}
      value="Rachel Green"
    />
  </div>
</styled.div>
`;

exports[`FieldView render errors = [...] 1`] = `
<styled.div
  aria-label="Field"
  data-dirty={false}
  data-disabled={false}
  data-empty={false}
  data-invalid={false}
  data-required={false}
  id="firstName"
>
  <div
    aria-label="Header"
  >
    <div
      aria-label="Label"
    >
      First Name
    </div>
    <i
      aria-label="Description"
      title="Enter your first name"
    >
      ?
    </i>
  </div>
  <div
    aria-label="Component"
  >
    <mockConstructor
      dirty={false}
      disabled={false}
      empty={false}
      invalid={false}
      onStateChange={[MockFunction]}
      onValueChange={[MockFunction]}
      required={false}
      state={Object {}}
      value="Rachel Green"
    />
  </div>
  <div
    aria-label="Error"
  >
    min length is 2
  </div>
</styled.div>
`;

exports[`FieldView render excluded = true 1`] = `""`;

exports[`FieldView render id = undefined 1`] = `
<styled.div
  aria-label="Field"
  data-dirty={false}
  data-disabled={false}
  data-empty={false}
  data-invalid={false}
  data-required={false}
>
  <div
    aria-label="Header"
  >
    <div
      aria-label="Label"
    >
      First Name
    </div>
    <i
      aria-label="Description"
      title="Enter your first name"
    >
      ?
    </i>
  </div>
  <div
    aria-label="Component"
  >
    <mockConstructor
      dirty={false}
      disabled={false}
      empty={false}
      invalid={false}
      onStateChange={[MockFunction]}
      onValueChange={[MockFunction]}
      required={false}
      state={Object {}}
      value="Rachel Green"
    />
  </div>
</styled.div>
`;

exports[`FieldView render invalid = true 1`] = `
<styled.div
  aria-label="Field"
  data-dirty={false}
  data-disabled={false}
  data-empty={false}
  data-invalid={true}
  data-required={false}
  id="firstName"
>
  <div
    aria-label="Header"
  >
    <div
      aria-label="Label"
    >
      First Name
    </div>
    <i
      aria-label="Description"
      title="Enter your first name"
    >
      ?
    </i>
  </div>
  <div
    aria-label="Component"
  >
    <mockConstructor
      dirty={false}
      disabled={false}
      empty={false}
      invalid={true}
      onStateChange={[MockFunction]}
      onValueChange={[MockFunction]}
      required={false}
      state={Object {}}
      value="Rachel Green"
    />
  </div>
</styled.div>
`;

exports[`FieldView render label = undefined 1`] = `
<styled.div
  aria-label="Field"
  data-dirty={false}
  data-disabled={false}
  data-empty={false}
  data-invalid={false}
  data-required={false}
  id="firstName"
>
  <div
    aria-label="Component"
  >
    <mockConstructor
      dirty={false}
      disabled={false}
      empty={false}
      invalid={false}
      onStateChange={[MockFunction]}
      onValueChange={[MockFunction]}
      required={false}
      state={Object {}}
      value="Rachel Green"
    />
  </div>
</styled.div>
`;

exports[`FieldView render onStateChange = undefined 1`] = `
<styled.div
  aria-label="Field"
  data-dirty={false}
  data-disabled={false}
  data-empty={false}
  data-invalid={false}
  data-required={false}
  id="firstName"
>
  <div
    aria-label="Header"
  >
    <div
      aria-label="Label"
    >
      First Name
    </div>
    <i
      aria-label="Description"
      title="Enter your first name"
    >
      ?
    </i>
  </div>
  <div
    aria-label="Component"
  >
    <mockConstructor
      dirty={false}
      disabled={false}
      empty={false}
      invalid={false}
      onStateChange={[Function]}
      onValueChange={[MockFunction]}
      required={false}
      state={Object {}}
      value="Rachel Green"
    />
  </div>
</styled.div>
`;

exports[`FieldView render onValueChange = undefined 1`] = `
<styled.div
  aria-label="Field"
  data-dirty={false}
  data-disabled={false}
  data-empty={false}
  data-invalid={false}
  data-required={false}
  id="firstName"
>
  <div
    aria-label="Header"
  >
    <div
      aria-label="Label"
    >
      First Name
    </div>
    <i
      aria-label="Description"
      title="Enter your first name"
    >
      ?
    </i>
  </div>
  <div
    aria-label="Component"
  >
    <mockConstructor
      dirty={false}
      disabled={false}
      empty={false}
      invalid={false}
      onStateChange={[MockFunction]}
      onValueChange={[Function]}
      required={false}
      state={Object {}}
      value="Rachel Green"
    />
  </div>
</styled.div>
`;

exports[`FieldView render required = true 1`] = `
<styled.div
  aria-label="Field"
  data-dirty={false}
  data-disabled={false}
  data-empty={false}
  data-invalid={false}
  data-required={true}
  id="firstName"
>
  <div
    aria-label="Header"
  >
    <div
      aria-label="Label"
    >
      First Name
    </div>
    <div
      aria-label="Required"
    >
      *
    </div>
    <i
      aria-label="Description"
      title="Enter your first name"
    >
      ?
    </i>
  </div>
  <div
    aria-label="Component"
  >
    <mockConstructor
      dirty={false}
      disabled={false}
      empty={false}
      invalid={false}
      onStateChange={[MockFunction]}
      onValueChange={[MockFunction]}
      required={true}
      state={Object {}}
      value="Rachel Green"
    />
  </div>
</styled.div>
`;

exports[`FieldView render state = undefined 1`] = `
<styled.div
  aria-label="Field"
  data-dirty={false}
  data-disabled={false}
  data-empty={false}
  data-invalid={false}
  data-required={false}
  id="firstName"
>
  <div
    aria-label="Header"
  >
    <div
      aria-label="Label"
    >
      First Name
    </div>
    <i
      aria-label="Description"
      title="Enter your first name"
    >
      ?
    </i>
  </div>
  <div
    aria-label="Component"
  >
    <mockConstructor
      dirty={false}
      disabled={false}
      empty={false}
      invalid={false}
      onStateChange={[MockFunction]}
      onValueChange={[MockFunction]}
      required={false}
      state={Object {}}
      value="Rachel Green"
    />
  </div>
</styled.div>
`;

exports[`FieldView render un-supported cofi field props - pass it to the underline component 1`] = `
<styled.div
  aria-label="Field"
  data-dirty={false}
  data-disabled={false}
  data-empty={false}
  data-invalid={false}
  data-required={false}
  id="firstName"
>
  <div
    aria-label="Header"
  >
    <div
      aria-label="Label"
    >
      First Name
    </div>
    <i
      aria-label="Description"
      title="Enter your first name"
    >
      ?
    </i>
  </div>
  <div
    aria-label="Component"
  >
    <mockConstructor
      dirty={false}
      disabled={false}
      empty={false}
      invalid={false}
      mickey="mouse"
      onStateChange={[MockFunction]}
      onValueChange={[MockFunction]}
      required={false}
      state={Object {}}
      value="Rachel Green"
    />
  </div>
</styled.div>
`;

exports[`FieldView render value = undefined 1`] = `
<styled.div
  aria-label="Field"
  data-dirty={false}
  data-disabled={false}
  data-empty={false}
  data-invalid={false}
  data-required={false}
  id="firstName"
>
  <div
    aria-label="Header"
  >
    <div
      aria-label="Label"
    >
      First Name
    </div>
    <i
      aria-label="Description"
      title="Enter your first name"
    >
      ?
    </i>
  </div>
  <div
    aria-label="Component"
  >
    <mockConstructor
      dirty={false}
      disabled={false}
      empty={false}
      invalid={false}
      onStateChange={[MockFunction]}
      onValueChange={[MockFunction]}
      required={false}
      state={Object {}}
    />
  </div>
</styled.div>
`;
